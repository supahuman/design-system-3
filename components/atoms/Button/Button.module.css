/* Button component styles using CSS Modules */

/* Variants */
.primary {
  /* Base button styles */
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  border: none;
  border-radius: var(--radii-base);
  font-family: var(--typography-fontFamily);
  font-weight: var(--typography-fontWeight-medium);
  cursor: pointer;
  transition:
    background-color 200ms ease,
    color 200ms ease,
    border-color 200ms ease,
    box-shadow 200ms ease,
    transform 150ms ease;
  text-decoration: none;
  outline: none;
  /* Primary variant styles */
  background-color: var(--colors-primary);
  color: white;
}

.primary:hover:not(:disabled) {
  background-color: color-mix(in srgb, var(--colors-primary) 85%, black);
}

.primary:focus-visible {
  box-shadow: 0 0 0 3px var(--colors-focus, rgba(59, 130, 246, 0.3));
}

.primary:active:not(:disabled) {
  transform: translateY(1px);
}

.primary:disabled {
  opacity: 0.5;
  cursor: not-allowed;
  transform: none;
}

.secondary {
  /* Base button styles */
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  border: none;
  border-radius: var(--radii-base);
  font-family: var(--typography-fontFamily);
  font-weight: var(--typography-fontWeight-medium);
  cursor: pointer;
  transition:
    background-color 200ms ease,
    color 200ms ease,
    border-color 200ms ease,
    box-shadow 200ms ease,
    transform 150ms ease;
  text-decoration: none;
  outline: none;
  /* Secondary variant styles */
  background-color: var(--colors-secondary);
  color: white;
}

.secondary:hover:not(:disabled) {
  background-color: color-mix(in srgb, var(--colors-secondary) 85%, black);
}

.secondary:focus-visible {
  box-shadow: 0 0 0 3px var(--colors-focus, rgba(59, 130, 246, 0.3));
}

.secondary:active:not(:disabled) {
  transform: translateY(1px);
}

.secondary:disabled {
  opacity: 0.5;
  cursor: not-allowed;
  transform: none;
}

/* Sizes */
.small {
  font-size: var(--typography-fontSize-sm);
  padding: var(--spacing-xs) var(--spacing-sm);
  min-height: 32px;
}

.medium {
  font-size: var(--typography-fontSize-base);
  padding: var(--spacing-sm) var(--spacing-md);
  min-height: 40px;
}

.large {
  font-size: var(--typography-fontSize-lg);
  padding: var(--spacing-md) var(--spacing-lg);
  min-height: 48px;
}

/* Loading state */
.loading {
  color: transparent;
}

.spinner {
  position: absolute;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.spinnerIcon {
  width: 1.25rem;
  height: 1.25rem;
  animation: spin 0.8s linear infinite;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
