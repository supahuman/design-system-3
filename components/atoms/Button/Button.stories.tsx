import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './Button.js';

const meta: Meta<typeof Button> = {
  title: 'Atoms/Button',
  component: Button,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A flexible button component with multiple variants and sizes. Built with CSS Modules for clean, maintainable styling.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: { type: 'select' },
      options: ['primary', 'secondary'],
      description: 'Visual variant of the button',
    },
    size: {
      control: { type: 'select' },
      options: ['small', 'medium', 'large'],
      description: 'Size of the button',
    },
    isLoading: {
      control: { type: 'boolean' },
      description: 'Whether the button is in a loading state',
    },
    disabled: {
      control: { type: 'boolean' },
      description: 'Whether the button is disabled',
    },
    children: {
      control: { type: 'text' },
      description: 'Content to display inside the button',
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

// Default button
export const Default: Story = {
  args: {
    children: 'Button',
  },
};

// Primary variant (default)
export const Primary: Story = {
  args: {
    variant: 'primary',
    children: 'Primary Button',
  },
};

// Secondary variant
export const Secondary: Story = {
  args: {
    variant: 'secondary',
    children: 'Secondary Button',
  },
};

// Different sizes
export const Sizes: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
      <Button size="small">Small</Button>
      <Button size="medium">Medium</Button>
      <Button size="large">Large</Button>
    </div>
  ),
};

// Loading states
export const Loading: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
      <Button isLoading variant="primary">
        Loading...
      </Button>
      <Button isLoading variant="secondary">
        Loading...
      </Button>
    </div>
  ),
};

// Disabled states
export const Disabled: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
      <Button disabled variant="primary">
        Disabled Primary
      </Button>
      <Button disabled variant="secondary">
        Disabled Secondary
      </Button>
    </div>
  ),
};

// All variants showcase
export const Showcase: Story = {
  render: () => (
    <div style={{ display: 'flex', flexDirection: 'column', gap: '2rem' }}>
      <div>
        <h3 style={{ marginBottom: '1rem', fontSize: '1.125rem', fontWeight: '600' }}>
          Primary Buttons
        </h3>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button variant="primary" size="small">
            Small Primary
          </Button>
          <Button variant="primary" size="medium">
            Medium Primary
          </Button>
          <Button variant="primary" size="large">
            Large Primary
          </Button>
        </div>
      </div>
      <div>
        <h3 style={{ marginBottom: '1rem', fontSize: '1.125rem', fontWeight: '600' }}>
          Secondary Buttons
        </h3>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button variant="secondary" size="small">
            Small Secondary
          </Button>
          <Button variant="secondary" size="medium">
            Medium Secondary
          </Button>
          <Button variant="secondary" size="large">
            Large Secondary
          </Button>
        </div>
      </div>
      <div>
        <h3 style={{ marginBottom: '1rem', fontSize: '1.125rem', fontWeight: '600' }}>States</h3>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button>Normal</Button>
          <Button disabled>Disabled</Button>
          <Button isLoading>Loading</Button>
        </div>
      </div>
    </div>
  ),
};
