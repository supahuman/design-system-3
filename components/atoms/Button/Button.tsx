import type { ButtonHTMLAttributes, ReactNode } from 'react';
import './Button.css';

/**
 * Button variants
 */
export type ButtonVariant = 'primary' | 'secondary' | 'ghost' | 'danger';

/**
 * Button sizes
 */
export type ButtonSize = 'small' | 'medium' | 'large';

/**
 * Button component props
 */
export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Content to display inside the button */
  children: ReactNode;
  /** Visual variant of the button */
  variant?: ButtonVariant;
  /** Size of the button */
  size?: ButtonSize;
  /** Whether the button is in a loading state */
  isLoading?: boolean;
  /** Icon to display before the button text */
  leftIcon?: ReactNode;
  /** Icon to display after the button text */
  rightIcon?: ReactNode;
  /** Additional CSS class */
  className?: string;
}

/**
 * Button component - a flexible button with multiple variants and sizes
 */
export function Button({
  children,
  variant = 'primary',
  size = 'medium',
  isLoading = false,
  leftIcon,
  rightIcon,
  className = '',
  disabled = false,
  ...props
}: ButtonProps) {
  // Build class names using ES6+ template literals
  const buttonClass = `ds-button ds-button-${variant} ds-button-${size} ${className} ${
    isLoading ? 'ds-button-loading' : ''
  }`.trim();

  return (
    <button className={buttonClass} disabled={disabled || isLoading} {...props}>
      {isLoading && <span className="ds-button-spinner" />}
      {leftIcon && <span className="ds-button-icon ds-button-icon-left">{leftIcon}</span>}
      <span className="ds-button-text">{children}</span>
      {rightIcon && <span className="ds-button-icon ds-button-icon-right">{rightIcon}</span>}
    </button>
  );
}
