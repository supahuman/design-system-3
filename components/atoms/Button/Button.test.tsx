import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Button } from './Button.js';

describe('Button', () => {
  it('renders with default props', () => {
    render(<Button>Click me</Button>);

    const button = screen.getByRole('button', { name: 'Click me' });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('primary', 'medium');
  });

  it('renders with different variants', () => {
    const { rerender } = render(<Button variant="primary">Primary</Button>);
    expect(screen.getByRole('button')).toHaveClass('primary');

    rerender(<Button variant="secondary">Secondary</Button>);
    expect(screen.getByRole('button')).toHaveClass('secondary');
  });

  it('renders with different sizes', () => {
    const { rerender } = render(<Button size="small">Small</Button>);
    expect(screen.getByRole('button')).toHaveClass('small');

    rerender(<Button size="medium">Medium</Button>);
    expect(screen.getByRole('button')).toHaveClass('medium');

    rerender(<Button size="large">Large</Button>);
    expect(screen.getByRole('button')).toHaveClass('large');
  });

  it('handles click events', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me</Button>);

    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('is disabled when disabled prop is true', () => {
    render(<Button disabled>Disabled</Button>);

    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
  });

  it('is disabled when loading', () => {
    render(<Button isLoading>Loading</Button>);

    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    expect(button).toHaveClass('loading');
  });

  it('shows loading spinner when isLoading is true', () => {
    render(<Button isLoading>Loading</Button>);

    const spinner = screen.getByRole('button').querySelector('.spinner');
    expect(spinner).toBeInTheDocument();

    const spinnerIcon = spinner?.querySelector('.spinnerIcon');
    expect(spinnerIcon).toBeInTheDocument();
  });

  it('does not call onClick when disabled', () => {
    const handleClick = jest.fn();
    render(
      <Button disabled onClick={handleClick}>
        Disabled
      </Button>,
    );

    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('does not call onClick when loading', () => {
    const handleClick = jest.fn();
    render(
      <Button isLoading onClick={handleClick}>
        Loading
      </Button>,
    );

    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('renders with left icon', () => {
    const leftIcon = <span data-testid="left-icon">ðŸ‘ˆ</span>;
    render(<Button leftIcon={leftIcon}>With Left Icon</Button>);

    expect(screen.getByTestId('left-icon')).toBeInTheDocument();
    expect(screen.getByText('With Left Icon')).toBeInTheDocument();
  });

  it('renders with right icon', () => {
    const rightIcon = <span data-testid="right-icon">ðŸ‘‰</span>;
    render(<Button rightIcon={rightIcon}>With Right Icon</Button>);

    expect(screen.getByTestId('right-icon')).toBeInTheDocument();
    expect(screen.getByText('With Right Icon')).toBeInTheDocument();
  });

  it('hides icons when loading', () => {
    const leftIcon = <span data-testid="left-icon">ðŸ‘ˆ</span>;
    const rightIcon = <span data-testid="right-icon">ðŸ‘‰</span>;

    render(
      <Button isLoading leftIcon={leftIcon} rightIcon={rightIcon}>
        Loading
      </Button>,
    );

    expect(screen.queryByTestId('left-icon')).not.toBeInTheDocument();
    expect(screen.queryByTestId('right-icon')).not.toBeInTheDocument();
  });

  it('applies custom className', () => {
    render(<Button className="custom-class">Custom</Button>);

    expect(screen.getByRole('button')).toHaveClass('custom-class');
  });

  it('forwards additional props to button element', () => {
    render(
      <Button data-testid="custom-button" aria-label="Custom label">
        Button
      </Button>,
    );

    const button = screen.getByTestId('custom-button');
    expect(button).toHaveAttribute('aria-label', 'Custom label');
  });

  it('has proper accessibility attributes', () => {
    render(<Button>Accessible Button</Button>);

    const button = screen.getByRole('button', { name: 'Accessible Button' });
    expect(button).toBeInTheDocument();
  });

  it('maintains focus management', () => {
    render(<Button>Focusable</Button>);

    const button = screen.getByRole('button');
    button.focus();
    expect(button).toHaveFocus();
  });

  describe('Snapshot Tests', () => {
    it('matches snapshot for primary variant', () => {
      const { container } = render(<Button variant="primary">Primary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for secondary variant', () => {
      const { container } = render(<Button variant="secondary">Secondary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for loading state', () => {
      const { container } = render(<Button isLoading>Loading Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for disabled state', () => {
      const { container } = render(<Button disabled>Disabled Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
