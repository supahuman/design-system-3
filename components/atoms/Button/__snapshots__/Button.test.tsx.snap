// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Snapshot Tests matches snapshot for disabled state 1`] = `
<button
  class="primary medium"
  disabled=""
>
  <span>
    Disabled Button
  </span>
</button>
`;

exports[`Button Snapshot Tests matches snapshot for loading state 1`] = `
<button
  class="primary medium loading"
  disabled=""
>
  <span>
    Loading Button
  </span>
  <div
    class="spinner"
  >
    <svg
      class="spinnerIcon"
      fill="none"
      viewBox="0 0 24 24"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
  </div>
</button>
`;

exports[`Button Snapshot Tests matches snapshot for primary variant 1`] = `
<button
  class="primary medium"
>
  <span>
    Primary Button
  </span>
</button>
`;

exports[`Button Snapshot Tests matches snapshot for secondary variant 1`] = `
<button
  class="secondary medium"
>
  <span>
    Secondary Button
  </span>
</button>
`;
